### DISCUSSION TITLE
property case convention

### DISCUSSION THREAD

**Discussion by squareborg** [spatie/laravel-data#21] [General]
Hey, Love the package. 

Does anyone have any ways to deal with case? Our endpoints generally use snake_case for input and output, 

So this makes us have to use `snake_case` properties in the data

```
class FooData extends Data
{
    public function __construct(
        public int $foo_id,
        public int $total_bar,
    ) {
    }
}
```

So now when using the data we access it via `$data->total_bar` and it makes us have inconsistent naming convention when most of our properties are `camelCase`.

Does anyone have a strategy for dealing with this?

Thanks

**Comment by rubenvanassche** [comment]
I'm not against using different conventions for property names since they make it very clear a property is part of a data object.

A solution to this problem could be to overwrite the `__get` and `__set` methods of your data object, you could then convert the name from camelcase to snake case. But then you would lose completions within your IDE for the camelcased properties.

**Comment by mpskovvang** [comment]
I'm in the same situation. API resources and database (Eloquent models) use snake case, but I prefer camel case for properties.

I ended up overriding the `from` and `transform` methods:

```PHP
class Address extends Data
{
    public function __construct(
        public string $addressLine1,
        public string|null $addressLine2 = null,
        public string $locality,
        public string $postalCode,
        public string $countryCode,
    ) {}

    public static function from($payload): static
    {
        $payload = [
            'addressLine1' => $payload['address_line_1'],
            'addressLine2' => $payload['address_line_2'],
            'locality' => $payload['locality'],
            'postalCode' => $payload['postal_code'],
            'countryCode' => $payload['country_code'],
        ];

        return parent::from($payload);
    }

    public function transform(TransformationType $type): array
    {
        $payload = parent::transform($type);

        return [
            'address_line_1' => $payload['addressLine1'],
            'address_line_2' => $payload['addressLine2'],
            'locality' => $payload['locality'],
            'postal_code' => $payload['postalCode'],
            'country_code' => $payload['countryCode'],
        ];
    }
}
```

This could probably be simplified with a custom snake case `Transformer` and `DataFromArrayResolver` using the `Str::camel` and `Str::snake` helpers. 

Would you recommend that approach? 🙂 

---

On the other hand: The naming convention for Eloquent model attributes is already snake case. Perhaps this is just something to get used to.

**Reply by rubenvanassche** [reply]
That's indeed a good solution, you could spare some work and keystrokes by creating a Data class within your project extending from the package Data class. In here you can overwrite the `from` and `transform` properties so that each payload with properties coming in is automatically transformed to the correct casing.
